### PR TITLE
Add ORTModelXXX for audio

### PR DESCRIPTION
**Pull request by mht-sharma**
# What does this PR do?

Add `ORTModel` for audio models inference

The PR adds:

- `ORTModelForAudioClassification`: ORT inference for [AutoModelForAudioClassification](https://huggingface.co/docs/transformers/model_doc/auto#transformers.AutoModelForAudioClassification) model
- `ORTModelForCTC`: ORT inference for [AutoModelForCTC](https://huggingface.co/docs/transformers/model_doc/auto#transformers.AutoModelForCTC) model
- `ORTModelForAudioXVector`: ORT inference for [AutoModelForAudioXVector](https://huggingface.co/docs/transformers/model_doc/auto#transformers.AutoModelForAudioXVector) model
- `ORTModelForAudioFrameClassification`: ORT inference for [AutoModelForAudioFrameClassification](https://huggingface.co/docs/transformers/model_doc/auto#transformers.AutoModelForAudioFrameClassification) model

Updates
- `ORTModelForSpeechSeq2Seq`: ORT inference for [AutoModelForSpeechSeq2Seq](https://huggingface.co/docs/transformers/model_doc/auto#transformers.AutoModelForSpeechSeq2Seq) model

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

### CODE REVIEW

File: optimum/onnxruntime/modeling_ort.py [hunk 1]
```diff
@@ -117,6 +124,13 @@
             Pixel values can be obtained from encoded images using [`AutoFeatureExtractor`](https://huggingface.co/docs/transformers/autoclass_tutorial#autofeatureextractor).
 """
 
+ONNX_AUDIO_INPUTS_DOCSTRING = r"""
+    Args:
+        input_values (`torch.Tensor` of shape `({0})`):
+            Float values of input raw speech waveform..
+            Input values can be obtained from audio file loaded into an array using [`AutoFeatureExtractor`](https://huggingface.co/docs/transformers/autoclass_tutorial#autofeatureextractor).
```
Comment (michaelbenayoun) [inline optimum/onnxruntime/modeling_ort.py:131]: Will the link work?

**Review by michaelbenayoun — APPROVED** [review]
LGTM!
Thanks @mht-sharma

File: optimum/onnxruntime/modeling_ort.py [hunk 2]
```diff
@@ -1545,6 +1559,341 @@ def _prepare_onnx_outputs(self, onnx_outputs):
         return outputs
 
 
+AUDIO_CLASSIFICATION_EXAMPLE = r"""
```
Comment (fxmarty) [inline optimum/onnxruntime/modeling_ort.py:1562]: Probably you already have - but can you make sure the code snippet works?

File: optimum/onnxruntime/modeling_ort.py [hunk 3]
```diff
@@ -1545,6 +1559,341 @@ def _prepare_onnx_outputs(self, onnx_outputs):
...
+    >>> onnx_ac = pipeline("audio-classification", model=model, feature_extractor=feature_extractor)
+
+    >>> pred = onnx_ac(dataset[0]["audio"]["array"])
+    ```
+"""
+
+
+@add_start_docstrings(
+    """
+    Onnx Model for sequence classification head on top (a linear layer over the pooled output) for tasks like
```
Comment (fxmarty) [inline optimum/onnxruntime/modeling_ort.py:1607]:
```suggestion
    Onnx Model with a sequence classification head on top (a linear layer over the pooled output) for tasks like
```

File: optimum/onnxruntime/modeling_ort.py [hunk 4]
```diff
@@ -1545,6 +1559,341 @@ def _prepare_onnx_outputs(self, onnx_outputs):
...
+
+            # run inference
+            outputs = self.model.run(None, onnx_inputs)
+            logits = torch.from_numpy(outputs[self.output_names["logits"]]).to(self.device)
+
+            # converts output to namedtuple for pipelines post-processing
+            return SequenceClassifierOutput(logits=logits)
+
+
+CTC_EXAMPLE = r"""
```
Comment (fxmarty) [inline optimum/onnxruntime/modeling_ort.py:1657]: same comment

File: optimum/onnxruntime/modeling_ort.py [hunk 5]
```diff
@@ -1545,6 +1559,341 @@ def _prepare_onnx_outputs(self, onnx_outputs):
...
+
+            # run inference
+            outputs = self.model.run(None, onnx_inputs)
+            logits = torch.from_numpy(outputs[self.output_names["logits"]]).to(self.device)
+
+            # converts output to namedtuple for pipelines post-processing
+            return CausalLMOutput(logits=logits)
+
+
+AUDIO_XVECTOR_EXAMPLE = r"""
```
Comment (fxmarty) [inline optimum/onnxruntime/modeling_ort.py:1733]: same

File: optimum/onnxruntime/__init__.py [hunk 6]
```diff
@@ -29,7 +29,11 @@
     ],
     "modeling_ort": [
         "ORTModel",
+        "ORTModelForAudioClassification",
+        "ORTModelForAudioFrameClassification",
+        "ORTModelForAudioXVector",
         "ORTModelForCustomTasks",
+        "ORTModelForCTC",
```
Comment (fxmarty) [inline optimum/onnxruntime/__init__.py:36]: Can you add them in the documentation .mdx as well?

File: optimum/onnxruntime/modeling_ort.py [hunk 7]
```diff
@@ -1545,6 +1559,341 @@ def _prepare_onnx_outputs(self, onnx_outputs):
...
+    ...     logits = model(**inputs).logits
+    >>> predicted_ids = torch.argmax(logits, dim=-1)
+
+    >>> transcription = processor.batch_decode(predicted_ids)
+"""
+
+
+@add_start_docstrings(
+    """
+    Onnx Model for with a `language modeling` head on top for Connectionist Temporal Classification (CTC).
```
Comment (fxmarty) [inline optimum/onnxruntime/modeling_ort.py:1685]:
```suggestion
    Onnx Model with a language modeling head on top for Connectionist Temporal Classification (CTC).
```

File: tests/onnxruntime/test_modeling.py [hunk 8]
```diff
@@ -2431,6 +2439,416 @@ def test_compare_to_io_binding(self, model_arch):
...
+        self.assertTrue("logits" in onnx_outputs)
+        self.assertIsInstance(onnx_outputs.logits, torch.Tensor)
+
+        with torch.no_grad():
+            transformers_outputs = transformers_model(**input_values)
+
+        # compare tensor outputs
+        self.assertTrue(torch.allclose(onnx_outputs.logits, transformers_outputs.logits, atol=1e-4))
+
+        gc.collect()
```
Comment (fxmarty) [inline tests/onnxruntime/test_modeling.py:2499]: We can leave it to an other PR (as you want), but following https://github.com/huggingface/optimum/pull/790 we will need to accept numpy inputs/outputs and change the tests accordingly.

File: tests/onnxruntime/test_modeling.py [hunk 9]
```diff
@@ -2431,6 +2439,416 @@ def test_compare_to_io_binding(self, model_arch):
...
+        self.assertIsInstance(outputs[0]["label"], str)
+
+    @parameterized.expand(
+        grid_parameters(
+            {"model_arch": SUPPORTED_ARCHITECTURES, "provider": ["CUDAExecutionProvider", "TensorrtExecutionProvider"]}
+        )
+    )
+    @require_torch_gpu
+    @pytest.mark.gpu_test
+    def test_pipeline_on_gpu(self, test_name: str, model_arch: str, provider: str):
```
Comment (fxmarty) [inline tests/onnxruntime/test_modeling.py:2539]: Could you make sure the tests pass on GPU?

**Review by fxmarty — APPROVED** [review]
Great PR, thank you! I left a few minor comments. As long as the tests (including GPU ones) pass, it is good to me.

**Comment by fxmarty** [comment]
> The PR adds:
>     ORTModelForAudioClassification
>     ORTModelForCTC
>     ORTModelForAudioXVector
>     ORTModelForAudioFrameClassification

Is there a reference in transformers doc what these correspond to? For the layman (like me), it can be tricky to know. Maybe it can be good to point in the doc (or docstrings of the class) a good reference for each in the transformers doc, if any exists.

File: tests/onnxruntime/test_modeling.py (hunk 8 above)
Comment (mht-sharma) [reply tests/onnxruntime/test_modeling.py]: will do in other PR

File: optimum/onnxruntime/__init__.py (hunk 6 above)
Comment (mht-sharma) [reply optimum/onnxruntime/__init__.py:36]: Done

File: optimum/onnxruntime/modeling_ort.py (hunk 2 above)
Comment (mht-sharma) [reply optimum/onnxruntime/modeling_ort.py:1665]: Done

File: optimum/onnxruntime/modeling_ort.py (hunk 1 above)
Comment (mht-sharma) [reply optimum/onnxruntime/modeling_ort.py:131]: yes

**Comment by mht-sharma** [comment]
> > The PR adds:
> > ORTModelForAudioClassification
> > ORTModelForCTC
> > ORTModelForAudioXVector
> > ORTModelForAudioFrameClassification
> 
> Is there a reference in transformers doc what these correspond to? For the layman (like me), it can be tricky to know. Maybe it can be good to point in the doc (or docstrings of the class) a good reference for each in the transformers doc, if any exists.

Added reference to `AutoModel`

> > The PR adds:
> > ORTModelForAudioClassification
> > ORTModelForCTC
> > ORTModelForAudioXVector
> > ORTModelForAudioFrameClassification
> 
> Is there a reference in transformers doc what these correspond to? For the layman (like me), it can be tricky to know. Maybe it can be good to point in the doc (or docstrings of the class) a good reference for each in the transformers doc, if any exists.

Added reference to `AutoModelXXX`

File: docs/source/onnxruntime/package_reference/modeling_ort.mdx [hunk 10]
```diff
@@ -66,3 +66,19 @@ specific language governing permissions and limitations under the License.
 ## ORTModelForVision2Seq
 
 [[autodoc]] onnxruntime.ORTModelForVision2Seq
+
+## ORTModelForAudioClassification
```
Comment (fxmarty) [inline docs/source/onnxruntime/package_reference/modeling_ort.mdx:70]: For now it was ordered alphabetically, but maybe it makes sense to have categories by modality (text, audio, image)

File: optimum/onnxruntime/modeling_ort.py [hunk 11]
```diff
@@ -1648,6 +1662,352 @@ def _prepare_onnx_inputs(self, use_torch: bool, **kwargs):
...
+    >>> if similarity < threshold:
+    ...     print("Speakers are not the same!")
+    >>> round(similarity.item(), 2)
+    ```
+"""
+
+
+@add_start_docstrings(
+    """
+    Onnx Model for with an XVector feature extraction head on top for tasks like Speaker Verification.
```
Comment (fxmarty) [inline optimum/onnxruntime/modeling_ort.py:1878]:
```suggestion
    Onnx Model with an XVector feature extraction head on top for tasks as speaker verification.
```

File: optimum/onnxruntime/modeling_seq2seq.py [hunk 12]
```diff
@@ -300,6 +307,12 @@ def forward(
             last_hidden_state = output_buffers["last_hidden_state"].view(output_shapes["last_hidden_state"])
         else:
             onnx_inputs = {"input_features": input_features.cpu().detach().numpy()}
+            if "attention_mask" in self.input_names:
+                onnx_inputs["attention_mask"] = attention_mask.cpu().detach().numpy()
+
+                # TODO: Replace with a better solution
+                if self.session.get_inputs()[1].type == "tensor(int64)":
+                    onnx_inputs["attention_mask"] = onnx_inputs["attention_mask"].astype(np.int64)
```
Comment (fxmarty) [inline optimum/onnxruntime/modeling_seq2seq.py:315]: Why is this needed?

File: optimum/onnxruntime/modeling_ort.py [hunk 13]
```diff
@@ -1648,6 +1662,352 @@ def _prepare_onnx_inputs(self, use_torch: bool, **kwargs):
...
+                input_values, ordered_input_names=self._ordered_input_names
+            )
+
+            # run inference with binding & synchronize in case of multiple CUDA streams
+            io_binding.synchronize_inputs()
+            self.model.run_with_iobinding(io_binding)
+            io_binding.synchronize_outputs()
+
+            # converts output to namedtuple for pipelines post-processing
+            return TokenClassifierOutput(logits=output_buffers["logits"].view(output_shapes["logits"]))
```
Comment (fxmarty) [inline optimum/onnxruntime/modeling_ort.py:1996]: So if I read correctly, this is not tested in test_modeling.py. Could we rather raise a `NotImplementedError` here?

File: tests/onnxruntime/test_modeling.py (hunk 8 above)
Comment (fxmarty) [reply tests/onnxruntime/test_modeling.py]: Why not in this PR? It shouldn't be much work.

File: optimum/onnxruntime/modeling_seq2seq.py (hunk 12 above)
Comment (mht-sharma) [reply optimum/onnxruntime/modeling_seq2seq.py]: Some of the model tokenisers generate input in `int32` but the model export always happens with `int64` dtype (int input generator always generates this input).  This cause an error during inference because of incorrect datatype.

File: optimum/onnxruntime/modeling_ort.py (hunk 13 above)
Comment (mht-sharma) [reply optimum/onnxruntime/modeling_ort.py:2029]: Sure

File: optimum/onnxruntime/modeling_seq2seq.py (hunk 12 above)
Comment (fxmarty) [reply optimum/onnxruntime/modeling_seq2seq.py]: Could you add a comment for this?

File: docs/source/onnxruntime/package_reference/modeling_ort.mdx (hunk 10 above)
Comment (mht-sharma) [reply docs/source/onnxruntime/package_reference/modeling_ort.mdx]: Added

File: optimum/onnxruntime/modeling_seq2seq.py (hunk 12 above)
Comment (mht-sharma) [reply optimum/onnxruntime/modeling_seq2seq.py]: Added

File: tests/onnxruntime/test_modeling.py (hunk 8 above)
Comment (mht-sharma) [reply tests/onnxruntime/test_modeling.py]: Done

Comment (mht-sharma) [reply tests/onnxruntime/test_modeling.py]: dcs
